src/drive/ducks/upload/index.js

```diff
@@ -120,3 +121,62 @@ export const getUploadQueue = state => state[SLUG].queue
 export const getProcessed = state =>
   getUploadQueue(state).filter(f => f.status !== PENDING)
 export const getSuccessful = state => getLoaded(getUploadQueue(state))
+
+// DnD helpers for folder upload
+export const canHandleFolders = evt => {
+  if (!evt.dataTransfer) return false
+  const dt = evt.dataTransfer
+  return dt.items && dt.items.length && dt.items[0].webkitGetAsEntry != null
+}
+
+const getFileFromEntry = async entry => {
+  let result
+  await new Promise(resolve => {
+    entry.file(file => {
+      result = file
+      resolve()
+    })
+  })
+  return result
+}
+
+const getFilesFromDirectory = async directory => {
+  let results = []
+  const dirReader = directory.createReader()
+  await new Promise(resolve => {
+    const entriesReader = async entries => {
+      for (let i = 0; i < entries.length; i += 1) {
+        const entry = entries[i]
+        if (entry.isFile) {
+          results.push(await getFileFromEntry(entry))
+        } else if (entry.isDirectory) {
+          results = results.concat(await getFilesFromDirectory(entry))
+        }
+      }
+      resolve()
+    }
+    dirReader.readEntries(entriesReader)
+  })
+  return results
+}
+
+export const getFilesFromItems = async evt => {
```

Comment:
It might be easier to test this function if it took the `dataTransfer' object directly instead of an event, wouldn't it?